324 fix manage account bug

machariamuguku opened the pull request:
Fixes #324 

- load logged in user data on clicking manage account
- 842513e194fa9b7c1a5db95d077c5024e9ced6a1 solves "a different version of babel-jest was detected higher up in the tree" issue when you clone and try to "lerna:prepublish" the repo for the first time
- e5e5cdc2ea3c9c58d92fcfdd8d8580c045efcfad exports abort controller from the fetch service class to use in useEffect cleanup functions with side effects (updating state)
- Prevent the edit and add user forms from re-render glitch (caused by props update after initial render)
- e15d9ca3852ce5fc3382d622efdb4c4d212fbfdb replaces the datasource array from the raw keycloakUsers to tableData which is filtered and has unique key props for each entry. solves the "Each child in an array should have a unique "key" prop" error when rendering a list of users

moshthepitt commented on packages/keycloak-user-management/src/components/CreateEditUser/index.tsx:
```diff
@@ -58,27 +59,46 @@ const CreateEditUser: React.FC<CreateEditPropTypes> = (props: CreateEditPropType
         .catch((_: Error) => sendErrorNotification(ERROR_OCCURED));
     }
+
+    return () => {
```
Does a test exist for this?

machariamuguku replied:
Not yet, let me work on that

machariamuguku replied:
Not yet, let me work on that

KipSigei reviewed (changes requested):
@machariamuguku we might need to add/update tests for these fixes

machariamuguku commented:
Updates:

- Fixed abort controller implementation in keycloak service class
- Added tests for aborting requests with abort controller
- 31e6c05521477c292d1c81d8324cf865ab7c3a94 removes the abort controller implementation from this branch for the following reasons: 

1. The abort signal had not been exported from the service class and thus any service method call does not return signal in its response. 
2. This change would have broken a lot of tests which assert on the response object.
3. Updating these tests would have been out of scope for this branch/issue.

- I took the initiative however, to work out the implementation and remove it in a single commit for future reference. 31e6c05521477c292d1c81d8324cf865ab7c3a94
- To fix the multiple form re-renders glitch issue, I added a loading status for transparency in conditional loading the spinner and the form. 575ad39a410fd5db7d46199f855b98df7ef577d9

machariamuguku commented:
> @machariamuguku we might need to add/update tests for these fixes

Done

p-netm commented on packages/keycloak-user-management/src/components/CreateEditUser/index.tsx:
```diff
@@ -62,72 +64,97 @@ const CreateEditUser: React.FC<CreateEditPropTypes> = (props: CreateEditPropType
         )
-        .catch((_: Error) => sendErrorNotification(lang.ERROR_OCCURED));
+        .catch((_: Error) => sendErrorNotification(lang.ERROR_OCCURED))
+        .finally(() => setLoading(false));
```
@machariamuguku looks like we are calling setLoading asynchronously in more than one  place, implicetely introduces a race condition that where we could be setting an incomplete loading state, should we avoid this?

machariamuguku replied:
The race condition is on setLoading which is overridden/skipped depending on if the other call is made with a similar or a different value. Assuming the first request sets Loading to true, and the second one attempts to set the same to true, the diffing algorithm takes charge and state is not updated (no state update, no rerender). Loading state is only updated if it's previous state is different than the one being set (in this case setLoading(true) before the first request and setLoading(false) after the last request)

machariamuguku replied:
Alternatively we could do `if (!loading) setLoading(true)` before requests and `if (loading) setLoading(false)` before requests?

machariamuguku replied:
Above confirmed by watching the loading state 
<img width="194" alt="Screenshot 2021-03-30 at 13 15 50" src="https://user-images.githubusercontent.com/10197807/112973728-4d9ab200-915a-11eb-907c-6a620fbdc9cd.png">

loading is only updated twice. First value is initial state, second is first update (to true), and third is last update (to false)
<img width="296" alt="Screenshot 2021-03-30 at 13 19 11" src="https://user-images.githubusercontent.com/10197807/112973961-8e92c680-915a-11eb-99ab-8207ca0f3b28.png">

p-netm replied:
I was thinking more on the case where you set loading to false while some promise is actually loading, i.e. one promise resolves before the other.

machariamuguku replied:
Any suggestions to avert that?
